au/code/au/utility/test/probable_primes_test.cc

```diff
@@ -0,0 +1,205 @@
+    // https://oeis.org/A001262
+    return {2047u,   3277u,   4033u,   4681u,   8321u,   15841u,  29341u,  42799u,
+            49141u,  52633u,  65281u,  74665u,  80581u,  85489u,  88357u,  90751u,
+            104653u, 130561u, 196093u, 220729u, 233017u, 252601u, 253241u, 256999u,
+            271951u, 280601u, 314821u, 357761u, 390937u, 458989u, 476971u, 486737u};
+}
+
+std::vector<uint64_t> miller_rabin_pseudoprimes_to_base_3() {
+    // https://oeis.org/A020229
+    return {121u,    703u,    1891u,   3281u,   8401u,   8911u,   10585u,  12403u,  16531u,
+            18721u,  19345u,  23521u,  31621u,  44287u,  47197u,  55969u,  63139u,  74593u,
+            79003u,  82513u,  87913u,  88573u,  97567u,  105163u, 111361u, 112141u, 148417u,
+            152551u, 182527u, 188191u, 211411u, 218791u, 221761u, 226801u};
+}
+
+TEST(MillerRabin, EvenNumbersAreBadInput) {
+    EXPECT_THAT(miller_rabin(2u, 0u), Eq(PrimeResult::BAD_INPUT));
+    EXPECT_THAT(miller_rabin(2u, 2u), Eq(PrimeResult::BAD_INPUT));
+    EXPECT_THAT(miller_rabin(2u, 4u), Eq(PrimeResult::BAD_INPUT));
+    EXPECT_THAT(miller_rabin(2u, 6u), Eq(PrimeResult::BAD_INPUT));
+    EXPECT_THAT(miller_rabin(2u, 8u), Eq(PrimeResult::BAD_INPUT));
+
+    EXPECT_THAT(miller_rabin(2u, 123456u), Eq(PrimeResult::BAD_INPUT));
+}
+
+TEST(MillerRabin, NumbersLessThanAPlusTwoAreBadInput) {
+    ASSERT_THAT(miller_rabin(9u, 11u), Eq(PrimeResult::PROBABLY_PRIME));
+
+    EXPECT_THAT(miller_rabin(10u, 11u), Eq(PrimeResult::BAD_INPUT));
+    EXPECT_THAT(miller_rabin(11u, 11u), Eq(PrimeResult::BAD_INPUT));
+}
+
+TEST(MillerRabin, MarksEveryPrimeAsProbablyPrime) {
+    auto expect_miller_rabin_probably_prime = [](std::size_t a, uint64_t n) {
+        const auto result = miller_rabin(a, n);
+        const auto expected =
+            (n < a + 2u || n % 2u == 0u) ? PrimeResult::BAD_INPUT : PrimeResult::PROBABLY_PRIME;
+
+        EXPECT_THAT(result, Eq(expected)) << "a = " << a << ", n = " << n;
+    };
+
+    const auto primes = first_n_primes<3'000u>();
+    for (const auto &p : primes) {
+        expect_miller_rabin_probably_prime(2u, p);
+        expect_miller_rabin_probably_prime(3u, p);
+        expect_miller_rabin_probably_prime(4u, p);
+        expect_miller_rabin_probably_prime(5u, p);
+
+        expect_miller_rabin_probably_prime(88u, p);
+    }
+}
+
+TEST(MillerRabin, OddNumberIsProbablyPrimeIffPrimeOrPseudoprime) {
+    const auto primes = first_n_primes<3'000u>();
+    const auto pseudoprimes = miller_rabin_pseudoprimes_to_base_2();
+
+    // Make sure that we are both _into the regime_ of the pseudoprimes, and that we aren't off the
+    // end of it.
+    ASSERT_THAT(primes.back(), AllOf(Gt(pseudoprimes.front()), Lt(pseudoprimes.back())));
+
+    std::size_t i_prime = 2u;  // Skip 2 and 3; they're too small for Miller-Rabin.
+    std::size_t i_pseudoprime = 0u;
+    for (uint64_t n = primes[i_prime]; i_prime < primes.size(); n += 2u) {
```

Comment:
Nit: Can we make `n` `const`?
Q: Why do we add by `2u` when  the primes are all odd after index 0, value 2?

Reply:
Sorry if it's a little confusing.

- We can't make `n` `const` because we're changing it as a loop variable.
- We're incrementing _`n`_ by two, not the _prime index_ by two.  `n` is how we _advance_ the loop, but `i_prime` is how we _end_ the loop: we're relying on `i_prime` being incremented internally.

It's a non-standard setup, but I think it's a good test: check every odd number until we run out of primes (having already made sure we run out of primes before pseudoprimes).

Reply:
Actually, this test is just needlessly confusing, and can become a lot more straightforward with a single simple tweak.